elytron/src/main/java/org/wildfly/extension/elytron/ElytronPersistentResourceXMLDescriptionFactory.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright The WildFly Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.wildfly.extension.elytron;
+
+import org.jboss.as.controller.AttributeMarshallers;
+import org.jboss.as.controller.AttributeParsers;
+import org.jboss.as.controller.PathElement;
+import org.jboss.as.controller.PersistentResourceXMLDescription;
+
+import java.util.function.Function;
+
+import static org.jboss.as.controller.PersistentResourceXMLDescription.builder;
+import static org.wildfly.extension.elytron.ElytronDescriptionConstants.DIR_CONTEXTS;
+import static org.wildfly.extension.elytron.ElytronDescriptionConstants.ENCRYPTION;
+import static org.wildfly.extension.elytron.ElytronDescriptionConstants.EXPRESSION;
+import static org.wildfly.extension.elytron.ElytronDescriptionConstants.EXPRESSION_RESOLVER;
+import static org.wildfly.extension.elytron.ElytronDescriptionConstants.PERMISSION_SETS;
+import static org.wildfly.extension.elytron.ElytronDescriptionConstants.SECURITY_PROPERTY;
+import static org.wildfly.extension.elytron.PermissionMapperDefinitions.PERMISSIONS;
+
+public enum ElytronPersistentResourceXMLDescriptionFactory implements Function<ElytronSubsystemSchema, PersistentResourceXMLDescription> {
+    INSTANCE;
+
+    @Override
+    public PersistentResourceXMLDescription apply(ElytronSubsystemSchema schema) {
+        PersistentResourceXMLDescription.PersistentResourceXMLBuilder builder = builder(ElytronExtension.SUBSYSTEM_PATH, schema.getNamespace());
+        if (schema.since(ElytronSubsystemSchema.VERSION_1_0)) {
+            builder.addAttribute(ElytronDefinition.DEFAULT_AUTHENTICATION_CONTEXT)
+                    .addAttribute(ElytronDefinition.INITIAL_PROVIDERS)
+                    .addAttribute(ElytronDefinition.FINAL_PROVIDERS)
+                    .addAttribute(ElytronDefinition.DISALLOWED_PROVIDERS)
+                    .addAttribute(ElytronDefinition.SECURITY_PROPERTIES, new AttributeParsers.PropertiesParser(null, SECURITY_PROPERTY, true), new AttributeMarshallers.PropertiesAttributeMarshaller(null, SECURITY_PROPERTY, true));
+        }
+
+        if (schema.since(ElytronSubsystemSchema.VERSION_5_0)) {
+            builder.addAttribute(ElytronDefinition.REGISTER_JASPI_FACTORY)
+                    .addAttribute(ElytronDefinition.DEFAULT_SSL_CONTEXT);
+        }
+
+        addAuthenticationClientParser(builder, schema);
+
+        if (schema.since(ElytronSubsystemSchema.VERSION_1_0)) {
```

Comment:
Just from a readability perspective, might be good to get this into an `addProviderParser` method like you have for the other cases so that in the future, any additional changes could be made in that method.

Reply:
Fixed